Segment Tree/LazyPropagation/README.markdown

```diff
@@ -0,0 +1,308 @@
+
+But that is common Segment Tree. By **Lazy Propagation**, we can implement to modify an interval in `O(logn)` time. Let's explore together in following:
+
+## `PushUp` - update to the top
+
+At first, we reference the implement of **Artur Antonov** about Segment Tree. This code contained *build*, *single update* and *interval query* three operation. The implement of *build* and *single update* operation is following:
+
+```swift
+public init(array: [T], leftBound: Int, rightBound: Int, function: @escaping (T, T) -> T) {
+	self.leftBound = leftBound
+	self.rightBound = rightBound
+	self.function = function
+	// ①
+	if leftBound == rightBound {
+		value = array[leftBound]
+	} 
+	// ②
+	else {
+		let middle = (leftBound + rightBound) / 2
+		leftChild = SegmentTree<T>(array: array, leftBound: leftBound, rightBound: middle, function: function)
+		rightChild = SegmentTree<T>(array: array, leftBound: middle+1, rightBound: rightBound, function: function)
+		value = function(leftChild!.value, rightChild!.value)
+	}
+}
+```
+
+In position ①, it means the current node is *leaf* because its left bound data is equal to the right. So we assign a value to it directly. In position ②, it means the current node is *parent* (which has one or more children), and we need to recursion down and update current node's data in the follow-up process.
+
+And then, we have a look for *interval query* operation:
+
+```swift
+public func query(leftBound: Int, rightBound: Int) -> T {
+	if self.leftBound == leftBound && self.rightBound == rightBound {
+		return self.value
+	}
+	guard let leftChild = leftChild else { fatalError("leftChild should not be nil") }
+	guard let rightChild = rightChild else { fatalError("rightChild should not be nil") }
+	// ①
+	if leftChild.rightBound < leftBound {
+		return rightChild.query(leftBound: leftBound, rightBound: rightBound)
+	} 
+	// ②
+	else if rightChild.leftBound > rightBound {
+		return leftChild.query(leftBound: leftBound, rightBound: rightBound)
+	}
+	// ③ 
+	else {
+		let leftResult = leftChild.query(leftBound: leftBound, rightBound: leftChild.rightBound)
+		let rightResult = rightChild.query(leftBound:rightChild.leftBound, rightBound: rightBound)
+		return function(leftResult, rightResult)
+	}
+}
+```
+
+Position ① means that the left bound of current query interval is on the  right of this right bound, so recurs to right direction. Position ② is opposite of position ①, recurs to left direction. Position ③ means our check interval is included the interval we need, so recurs deeply.
+
+![pushUp](Images/pushUp.png)
```

Comment:
This image is written by Chinese, please update it as English.

Reply:
Fix it!